File: DMOJ Solutions/ccc13s4.py
import sys
def main():
	n,m = map(int,input().split())
	graph = [set() for i in range(n+1)]
	for i in range(m):
		a,b = map(int,sys.stdin.readline().split())
		graph[a].add(b)
	p,q = map(int,input().split())


	#bfs
	from collections import deque
	d = deque()
	d.append(p)
	visited = [0 for i in range(n+1)]
	while d:
		node = d.pop()
		visited[node] = 1
		if node == q:
			return "yes"
		for neighbor in graph[node]:
			if not visited[neighbor]:
				d.append(neighbor)


	d = deque()
	d.append(q)
	visited = [0 for i in range(n+1)]
	while d:
		node = d.pop()
		visited[node] = 1
		if node == p:
			return "no"
		for neighbor in graph[node]:
			if not visited[neighbor]:
				d.append(neighbor)

	
	return "unknown"

print(main())

File: DMOJ Solutions/ccc13s4.cpp
#include <bits/stdc++.h>
using namespace std;

const int maxn = 1e6 + 5;
int n,m;
bool vis[maxn];
vector<int> adj[maxn];

bool bfs(int s, int e){
	queue<int> q;
	for (int i = 0; i < maxn; ++i) vis[i] = false;
	q.push(s);
	vis[s] = true;
	while (!q.empty()){
		int node = q.front(); q.pop();
		if (node == e) break;
		for (int v: adj[node]){
			if (!vis[v]){
				q.push(v);
				vis[v] = true;
			}
		}
	}
	return vis[e];
}

int32_t main(){
	cin.sync_with_stdio(0); cin.tie(0);
	cin >> n >> m;
	for (int i = 0; i < m; ++i){
		int x,y; cin >> x >> y;
		adj[x].emplace_back(y);
	}
	int p,q; cin >> p >> q;
	if (bfs(p,q)){
		cout << "yes" << "\n";
	}
	else if (bfs(q,p)){
		cout << "no" << "\n";
	}
	else {
		cout << "unknown" << "\n";
	}
	return 0;
}